modules/core/src/main/scala/doobie/util/read.scala

```diff
@@ -42,67 +44,120 @@ and similarly with Get:
+  def derived[A](implicit
+      @implicitNotFound(
+        "Cannot derive Read instance. Please check that each field in the case class has a Read instance or can derive one")
+      ev: Derived[MkRead[A]]
+  ): Read[A] = ev.instance.underlying
 
-  trait Auto {
-    implicit def deriveRead[A](implicit ev: MkRead[A]): Read[A] = ev
-  }
+  trait Auto extends MkReadInstances
 
   implicit val ReadApply: Applicative[Read] =
     new Applicative[Read] {
       def ap[A, B](ff: Read[A => B])(fa: Read[A]): Read[B] = fa.ap(ff)
-      def pure[A](x: A): Read[A] = new Read(Nil, (_, _) => x) {}
+      def pure[A](x: A): Read[A] = Read.Composite(Nil, _ => x)
       override def map[A, B](fa: Read[A])(f: A => B): Read[B] = fa.map(f)
     }
 
-  implicit val unit: Read[Unit] =
-    Read(Nil, (_, _) => ())
+  implicit val unit: Read[Unit] = Read.Composite(Nil, _ => ())
+
+  implicit val optionUnit: Read[Option[Unit]] = Read.Composite(Nil, _ => Some(()))
 
-  implicit val optionUnit: Read[Option[Unit]] =
-    Read(Nil, (_, _) => Some(()))
+  implicit def fromReadOption[A](implicit read: Read[A]): Read[Option[A]] = read.toOpt
 
-  implicit def fromGet[A](implicit ev: Get[A]): Read[A] =
-    new Read(List((ev, NoNulls)), ev.unsafeGetNonNullable) {}
+  /** Simple instance wrapping a Put. i.e. single column non-null value */
+  final case class Single[A](get: Get[A]) extends Read[A] {
+    def unsafeGet(rs: ResultSet, startIdx: Int): A =
+      get.unsafeGetNonNullable(rs, startIdx)
+
+    override def toOpt: Read[Option[A]] = SingleOpt(get)
+
+    override def gets: List[(Get[?], NullabilityKnown)] = List(get -> NoNulls)
+
+    override val length: Int = 1
+
+  }
+
+  /** Simple instance wrapping a Put. i.e. single column nullable value */
+  final case class SingleOpt[A](get: Get[A]) extends Read[Option[A]] {
+    def unsafeGet(rs: ResultSet, startIdx: Int): Option[A] =
+      get.unsafeGetNullable(rs, startIdx)
+
+    override def toOpt: Read[Option[Option[A]]] = Composite(List(this), l => Some(l.head.asInstanceOf[Option[A]]))
+    override def gets: List[(Get[?], NullabilityKnown)] = List(get -> Nullability.Nullable)
+
+    override val length: Int = 1
+  }
 
-  implicit def fromGetOption[A](implicit ev: Get[A]): Read[Option[A]] =
-    new Read(List((ev, Nullable)), ev.unsafeGetNullable) {}
+  /** A Read instance consists of multiple underlying Read instances */
+  final case class Composite[A](readInstances: List[Read[?]], construct: List[Any] => A) extends Read[A] {
+
+    override val length: Int = readInstances.map(_.length).sum
+
+    override lazy val gets: List[(Get[?], NullabilityKnown)] = readInstances.flatMap(_.gets)
+
+    override def unsafeGet(rs: ResultSet, startIdx: Int): A = {
+      import scala.collection.mutable
+      val accum = mutable.ArrayBuffer.empty[Any]
+      var idx = startIdx
+      readInstances.foreach { rr =>
+        accum += rr.unsafeGet(rs, idx)
+        idx += rr.length
+      }
+      construct(accum.toList)
```

Comment:
If `accum` is being turned into a list is it more efficient to use `ListBuffer` instead of `ArrayBuffer`?

Reply:
Nice - that seems to give a 20% boost in the benchmarks :)